### DISCUSSION TITLE
Cannot modify an evaluated object originating in an imported file

### DISCUSSION THREAD

**Discussion by netcorefan1** [NuGet/Home#12426] [Q&A] [answered]
Hello to all,
I need to create an unified local NuGet package which contains all the NuGet packages in order to be easily consumed from my local solutions.
This is my configuration:
```
<!-- SharedPackage.csproj -->
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net7.0</TargetFramework>
    <GeneratePackageOnBuild>True</GeneratePackageOnBuild>
  </PropertyGroup>
  <ItemGroup>
    <None Include="SharedPackage.props" Pack="true" PackagePath="build" />
  </ItemGroup>
</Project>

<!-- SharedPackage.props -->
<Project>
	<ItemGroup>
		<PackageReference Include="SuperLinq" Version="*-*" />
		<PackageReference Include="Serilog" Version="*-*"/>
	</ItemGroup>
</Project>
```
Installation works as expected:
![1](https://user-images.githubusercontent.com/63934725/217975844-b44b6cbf-bfd5-4732-9306-bf4cd4453349.jpg)
However, my main goal is the ability to uninstall the unused references when in release mode. VS correctly recognize unused ones, but when I start the uninstall process I get this error:
![2](https://user-images.githubusercontent.com/63934725/217975995-e675165c-a717-4add-ba71-1e7715497ab9.jpg)

Any help would be appreciated.
Thanks

**Comment by zivkan** [comment, marked as answer]
@netcorefan1 with PackageReference, packages are extracted to the "global packages folder" (`c:\users\username\.nuget\packages`) and all projects/solutions/repos on the machine will use that same package contents. Therefore, individual projects must not modify those files because 1. it means the files are no longer "correct" compared to the package (nupkg) contents, therefore if you build the same project on a different machine you will get a different outcome, and 2. it affects all projects on the machine, not just that one.

Secondly, VS has a performance optimization which breaks "correctness", so while VS will find the PackageReferences defined in your package's props file, command line restores will not (command line restore explicitly blocks importing props/targets from packges, so they can't change the restore graph.

Therefore, I strongly discourage you from trying to solve your problem in this way.

If you wish to build an [MSBuild SDK](https://learn.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk?view=vs-2022) that adds your common packages, that will work (MSBuild has a docs page on [how it builds projects](https://learn.microsoft.com/en-us/visualstudio/msbuild/build-process-overview?view=vs-2022), and if you understand this, then hopefully you'll understand why an SDK can add a PackageReference, but Restore running as a Target cannot). But Visual Studio's GUI still won't know how to remove the references. Anyone using your SDK will need to hand edit their csproj. Possibly by adding `<PackageReference Remove="PackageId" />`, or possibly doing something more complex. Since MSBuild is a programming language, there are many ways to do things, and therefore how to undo them depends on how it was done in the first place.

**Reply by netcorefan1** [reply]
Many thanks for your detailed explanation. I should have clarified better that I'm the only one consumer of the packages and everything is done locally on the same machine. Does you discouragement still applies for this particular situation (supposing there is a way to disable the lock) ?

I have a main solution which contains all the main projects and another separate folder which I use for testing, prototypes and other temporary quick tests. Once some project reach a satisfying semi-wroking model, then it is approved and moved into the main solution for further development. In each project I need to have at hands something like 30-40 nuget packages, but I may end up to really consume half, just a couples ones or even none. The main purpose of so many packages is mainly intellisense. If intellisense shows up a particular function I'm looking for on that particular moment, then I known that the related package will be consumed from the project.
Basically, more than the installation of the packages themself, I need their metadata for intellisense. Once I reference a function of interest, then I can proceed to installation. All of this is mainly to avoid to end up with thousands of duplicates and work with a clean workspace.
The main shared package contains all the references included my custom extensions, classes etc. Is located in the main solution and other than being used by the solution itself, I need it automatically available on each new project.

Regarding the automatic installation (and removal of the unused packages), I thought to a VSIX extension, but now, after your response, I'm wondering if I should let handle every aspect of the entire process to the extension, or separate the tasks (leave part of the job to MSBuild sdk as recommended from you and the rest to the extension). Would you think would be better to create an option page in VS which stores a wiew model of all the needed packages along with the logic for installation/removal and let VS do the job when the target events occur? Does VS sdk exposes unused references management and other necessary functions I may need for this job?
Thanks

**Reply by zivkan** [reply]
I'm sorry, I didn't read your message carefully, so it may be what I'm about to write might have obviously mistakes, but given that your scenario appears to be very unique to your workflow, I really need to focus on work that will benefit many customers.

Firstly, Visual Studio's Intellisense already has a database of types from the top X most downloaded packages from nuget.org. So, if the type you want to use is from a package on nuget.org, you don't need your own list of 30+ pages for Intellisense to work. Although, you will need to type the full type name once, then use the lightbulb to install the package. You also need nuget.org as a package source. If you use a private feed that upstreams to nuget.org, Intellisense doesn't know that nuget.org packages are available, so won't propose the package.

Additionally, in Visual Studio, Intellisense knows about types in packages used by other projects, so if project A uses a package that project B doesn't use, if you try to use the type from the package, the lightbulb fix will propose installing the package in project B for you, even if the package is not from nuget.org. So, it's not quite as convenient as it's nuget.org package database, 

It's not clear to me, from my quick skim over your message, if this is sufficient for what you're trying to achieve.

Otherwise, it seems to me like you might get "close enough" by having a `Directory.Build.props` in your solution root, rather than trying to use a package. You did mention solution, so I assume that this could work for you. You could also have a Directory.Build.props on your `c:\` drive root (or whatever drive you do your development on). But if any of your repos have their own Directory.Build.props files, then they'll explicitly need to import higher up D.B.p files, as per MSBuild's documentation. Usually this is not desirable for build security (importing files from outside the repo root), so that's something to consider.

As for the rest of the "polish" you're trying to get from a more sophisticated solution, is the benefit you believe you'll get greater than the effort of implementing it? The answer may be yes, I didn't really understand from your message what problem you're trying to solve with this, other than having types available in Intellisense.

In any case, NuGet just asks Visual Studio's project system to modify the project when packages are installed/upgraded/uninstalled, and it's up to the project system to determine which file to modify. It's actually the component throwing that exception about "cannot modify evaluated ...", not NuGet itself. I don't know when it throws that exception and when it allows modification. Technically the files in your global packages folder are modifyable by your user account (not read only), so I'm not sure how it decided to block modification of the file in the directory. Therefore, I don't know what other options will work and which will fail with the same error.

**Reply by netcorefan1** [reply]
Thanks again for your support. Don't worry, it has been only my fault.
Yes, I know about intellisense type suggestions, but, as you say, I need to know the kind of type in advance which is something that I don't know. The convenience of having a preselected list of installed packages is the fact that I can find what I need for directly from intellisense by reading xml docs and by typying a matching keyword. This is especially true in the case of extension methods. I realized that I could save a good amount of time if I can find what I need for without leaving VS by including a predefined list of nuget packages that can cover a good amount of use cases.

Your suggestion to use `Directory.Build.props` seems to be a good starting point. I renamed `SharedPackage.props` into `Directory.Build.props` and placed it into the drive root. References are automatically installed on each project. However, unfortunately, I'm facing the same lock issue and I can't find a way to uninstall a package. I even tried to place the file into `C:\Users\User\source\repos` thinking that could be a matter of unsufficient privileges, but same problem. This is just the default repo created by VS on first install on the user account, so I have no idea why installed packages are still marked with the readonly icon while VS recognize them as uninstallable: 
![3](https://user-images.githubusercontent.com/63934725/218234959-8a09b81c-1bde-4e81-b900-30c85892b26f.jpg). 
All packages comes from NuGet.org source and are stored in the "global packages folder" (`c:\users\username\.nuget\packages`).
Am I missing something?

To be honest, I asked myself if the efforts required to keep a cleaner workspace are justified, but after having already spent time on this it has become more of a challenge than else especially because it is hard to give up when the goal is simply to uninstall uneeded packages.

I will try your suggestion to play with MSBuild SDK to see if I can come with something closer to the goal.

**Reply by netcorefan1** [reply]
**Just some updates.** 
I played with the MSBuild SDK and I'm not sure is the right way to go. There is to learn its language model and go very deep in the exploration of the build process. It increase complexity and I have the impression that I could end up with the same problem. Even if I come up with something working, some tasks will still need to be performed from VS extensibility. I think I'm better off to do everything through the VS SDK for which I have the required knowledge. It should be pretty straightforward to interact with Nuget thanks to the exposed [API](https://learn.microsoft.com/en-us/nuget/visual-studio-extensibility/nuget-api-in-visual-studio). In this way I can have control on both Nuget and component. Since what I'm trying to can be manually done by hand, I can expect that it can be fully automated through VS. It will also make possible to easily change packages list or the logic for install/uninstall each time I need through the extension settings without even recompiling.

I don't want to further abuse your time, but I would really appreciate if you could check why the `Directory.Build.props` solution fails with the same lock problem because I start to think that there could be some problem with my machine.
Thanks

**Reply by zivkan** [reply]
>  I would really appreciate if you could check why the Directory.Build.props solution fails with the same lock problem because I start to think that there could be some problem with my machine.

NuGet does not modify project files (including *.props and *.targets) directly in Visual Studio. Instead, NuGet calls APIs on the project system, which is responsible for making changes to those MSBuild files. The error message you see is being propogated up from the project system. So, I don't know.

If you report a problem via Visual Studio's feedback system, we can assign the issue to the project system team, who can investigate, and reply if it's by design. If you're using SDK style projects, you could possibly ask https://github.com/dotnet/project-system (GitHub doesn't allow transfers of issues across different github orgs), but if you're using a non-SDK project, that's closed source and not on github, so the VS feedback system (developercommunity.visualstudio.com) is the only way to communicate with that team.

**Comment by netcorefan1** [comment]
Thanks for your help. As of now, *.props and *.targets have the expected behaviour for the other tasks I need to perform. I expect to not encounter such readonly problems when working from within a VSPackage, but If I do I will report this issue to the project system repo.
Thanks
